waspc/src/Analyzer/Parser/Lexer.x

```diff
@@ -119,21 +126,27 @@ lexer parseToken = do
 -- | Takes a lexeme like "{=json" and sets the quoter start code
 beginQuoter :: String -> Parser Token
 beginQuoter leftQuoteTag = do
-  setStartCode quoter
   let tag = drop 2 leftQuoteTag
+  setStartCode $ QuoterStartCode tag
   createConstToken (TLQuote tag) leftQuoteTag
 
--- | Takes a lexeme like "json=}" and returns to start code 0
-endQuoter :: String -> Parser Token
-endQuoter rightQuoteTag = do
-  setStartCode 0
-  let tag = take (length rightQuoteTag - 2) rightQuoteTag
-  createConstToken (TRQuote tag) rightQuoteTag
+-- | Takes a lexeme like "json=}" and either ends a quoter or add quoted text to the quoter
+lexQuoterEndTag :: String -> Parser Token
+lexQuoterEndTag rightQuoteTag = gets parserLexerStartCode >>= \startCode -> case startCode of
+  DefaultStartCode -> do
+    createConstToken (TRQuote tag) rightQuoteTag
+  QuoterStartCode startTag | startTag == tag -> do
+    setStartCode DefaultStartCode
+    createConstToken (TRQuote tag) rightQuoteTag
+  _ -> do
+    createValueToken TQuoted rightQuoteTag
+  where
+    tag = take (length rightQuoteTag - 2) rightQuoteTag
```

Comment:
After I read this function, I feel it is somewhat complicated, all the state switching. Well, it is complciated because of the state, and even worse, state is kind of "global" in the shape of the lexer start codes. But ok, I don't think we can do much about that, so I think it is ok. Comment above this function certainly helped me.

p.s. Regarding our discussion on resolving stuff, this is an example of comment where responsibility of resolving it falls onto you, unless you decide to make a reply in which case it falls onto me.